internal/statestore/redis.go

```diff
@@ -56,65 +56,126 @@ func (rb *redisBackend) Close() error {
+		Wait:         true,
+		TestOnBorrow: testOnBorrow,
+		DialContext:  getDialFunc(poolType, cfg, sentinelPool),
+	}
+}
 
-			_, err := c.Do("PING")
-			return err
-		},
-		DialContext: func(ctx context.Context) (redis.Conn, error) {
-			if ctx.Err() != nil {
-				return nil, ctx.Err()
+func getDialFunc(poolType string, cfg config.View, sentinelPool *redis.Pool) func(ctx context.Context) (redis.Conn, error) {
+	idleTimeout := cfg.GetDuration("redis.pool.idleTimeout")
+	healthCheckTimeout := cfg.GetDuration("redis.pool.healthCheckTimeout")
+
+	if cfg.GetBool("redis.sentinelEnabled") {
+		sentinelAddr := fmt.Sprintf("%s:%s", cfg.GetString("redis.sentinelHostname"), cfg.GetString("redis.sentinelPort"))
+		sentinelURL := redisURLFromAddr(sentinelAddr, cfg)
+
+		switch poolType {
+		case "healthcheck":
+			return func(ctx context.Context) (redis.Conn, error) {
+				if ctx.Err() != nil {
+					return nil, ctx.Err()
+				}
+				return redis.DialURL(sentinelURL, redis.DialConnectTimeout(healthCheckTimeout), redis.DialReadTimeout(healthCheckTimeout))
 			}
-			return redis.DialURL(redisURL, redis.DialConnectTimeout(cfg.GetDuration("redis.pool.idleTimeout")), redis.DialReadTimeout(cfg.GetDuration("redis.pool.idleTimeout")))
-		},
-	}
-	healthCheckPool := &redis.Pool{
-		MaxIdle:     3,
-		MaxActive:   0,
-		IdleTimeout: 10 * cfg.GetDuration("redis.pool.healthCheckTimeout"),
-		Wait:        true,
-		DialContext: func(ctx context.Context) (redis.Conn, error) {
-			if ctx.Err() != nil {
-				return nil, ctx.Err()
+		case "master":
+			return func(ctx context.Context) (redis.Conn, error) {
+				if ctx.Err() != nil {
+					return nil, ctx.Err()
+				}
+
+				sentinelConn, err := sentinelPool.GetContext(ctx)
+				if err != nil {
+					redisLogger.WithFields(logrus.Fields{
+						"error": err.Error(),
+					}).Error("failed to connect to redis sentinel")
+					return nil, status.Errorf(codes.Unavailable, "%v", err)
+				}
+
+				masterInfo, err := redis.Strings(sentinelConn.Do("SENTINEL", "GET-MASTER-ADDR-BY-NAME", cfg.GetString("redis.sentinelMaster")))
+				if err != nil {
+					redisLogger.WithFields(logrus.Fields{
+						"error": err.Error(),
+					}).Error("failed to get current master from redis sentinel")
+					return nil, status.Errorf(codes.Unavailable, "%v", err)
+				}
+
+				masterURL := redisURLFromAddr(fmt.Sprintf("%s:%s", masterInfo[0], masterInfo[1]), cfg)
+				return redis.DialURL(masterURL, redis.DialConnectTimeout(idleTimeout), redis.DialReadTimeout(idleTimeout))
 			}
-			return redis.DialURL(redisURL, redis.DialConnectTimeout(cfg.GetDuration("redis.pool.healthCheckTimeout")), redis.DialReadTimeout(cfg.GetDuration("redis.pool.healthCheckTimeout")))
-		},
-	}
-
-	return &redisBackend{
-		healthCheckPool: healthCheckPool,
-		redisPool:       pool,
-		cfg:             cfg,
+		default:
+			redisLogger.Fatalf("dial type: %s not supported, sentinelEnabled: %v", poolType, cfg.GetBool("redis.sentinelEnabled"))
+			return nil
+		}
+	} else {
```

Comment:
The more i read this, the more i think it might be readable to split out healthcheck / master pool fetch logic up top - and each can then have if sentinel enabled blocks within.